sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/TestMethodShouldNotBeIgnored.cs

```diff
@@ -51,46 +53,56 @@ public sealed class TestMethodShouldNotBeIgnored : SonarDiagnosticAnalyzer
-                    {
-                        return;
-                    }
-
-                    var attributes = methodSymbol.GetAttributes();
-                    if (!attributes.Any(a => a.AttributeClass.IsAny(TrackedTestMethodAttributes)))
-                    {
-                        return;
-                    }
-
-                    var ignoreAttribute = attributes.FirstOrDefault(a =>
-                        a.AttributeClass.Is(KnownType.Microsoft_VisualStudio_TestTools_UnitTesting_IgnoreAttribute));
-                    if (ignoreAttribute == null)
-                    {
-                        return;
-                    }
-
-                    var ignoreAttributeSyntax = ignoreAttribute.ApplicationSyntaxReference.GetSyntax();
-
-                    var hasWorkItemAttribute = attributes.Any(a =>
-                        a.AttributeClass.Is(KnownType.Microsoft_VisualStudio_TestTools_UnitTesting_WorkItemAttribute));
-                    var hasTrailingComment = ignoreAttributeSyntax.Parent.GetTrailingTrivia()
-                        .Any(trivia => trivia.IsKind(SyntaxKind.SingleLineCommentTrivia));
-
-                    if (!hasWorkItemAttribute &&
-                        !hasTrailingComment)
-                    {
-                        c.ReportDiagnosticWhenActive(Diagnostic.Create(rule, ignoreAttributeSyntax.GetLocation()));
-                    }
-                }, SyntaxKind.MethodDeclaration);
-        }
+            context.RegisterSyntaxNodeActionInNonGenerated(c =>
+            {
+                var attribute = (AttributeSyntax)c.Node;
+                if (HasReasonPhrase(attribute) ||
+                    HasTrailingComment(attribute) ||
+                    !IsMsTestIgnoreAttribute(attribute, c.SemanticModel))
+                {
+                    return;
+                }
+
+                var attributeTarget = attribute.Parent?.Parent;
+                if (attributeTarget == null)
+                {
+                    return;
+                }
+
+                var testMethodOrClass = c.SemanticModel.GetDeclaredSymbol(attributeTarget);
+
+                if (testMethodOrClass != null &&
+                    testMethodOrClass.GetAttributes().Any(IsTestOrTestClassAttribute) &&
+                    !testMethodOrClass.GetAttributes().Any(IsWorkItemAttribute))
+                {
+                    c.ReportDiagnosticWhenActive(Diagnostic.Create(rule, attribute.GetLocation()));
+                }
+            },
+            SyntaxKind.Attribute);
+        }
+
+        private bool HasReasonPhrase(AttributeSyntax ignoreAttributeSyntax) =>
+            ignoreAttributeSyntax.ArgumentList?.Arguments.Count > 0; // Any ctor argument counts are reason phrase
+
+        private static bool HasTrailingComment(SyntaxNode ignoreAttributeSyntax) =>
+            ignoreAttributeSyntax.Parent.GetTrailingTrivia()
+                .Any(trivia => trivia.IsKind(SyntaxKind.SingleLineCommentTrivia));

```

Comment:
I know this was like this but maybe we should also handle multiline comments `/* Ignored because something */`